### ISSUE TITLE
Customizing tooltip's contents *updated: Aug 1st*

### ISSUE THREAD

**Issue by veselinodjov** [swimlane/ngx-charts#932]
**I'm submitting a ...**  (check one with "x")

- [ ] bug report - search github for a similar issue or PR before submitting
- [ ] feature request
- [x] support request - use StackOverflow (add the `ngx-charts` tag) or the gitter chat for support questions

Hello everyone,

I've been trying to set a custom label in a custom line chart's tooltip , e.g., modified number of minutes in HH:mm format (77 min --> 1:17), for quite some time now, but unfortunately without any success. Displaying the value as 1.2833...(3) is not an alternative.

![float_to_hh_mm](https://user-images.githubusercontent.com/23047676/43405060-47939bc4-9419-11e8-9071-ff3f7e5a2cb9.png)

Does anybody know how to preserve the x and y axis values (a date and a number respectively), and modify the tooltip display value? 

For example: https://swimlane.github.io/ngx-charts/#/ngx-charts/line-chart

Instead of having a tooltip that displays:
###  Color, Country name and a String (e.g. Number > 3000 ? 'high' : 'low';)

<details>
  <summary>
    Show old issues (from beginning up to Jul 31st, incl.)
  </summary>

Current situation, Jul 31th: 

Adding the   [tooltipTemplate]="seriesTooltipTemplate"  to   ngx-charts-**circle-series** solved one of the problems and created another - minutes will be dispalyed as HH:mm via a custom pipe, but the name of the parameter is lost along the way.

![float_to_hh_mm_pipe_1](https://user-images.githubusercontent.com/23047676/43461344-91ecd8e2-94d3-11e8-896b-373a7a77ed87.png)

As pseudo code intended:

```
<ngx-my-chart 
   [options]="options_values"
   <svgs.../>
   
   <svg:g ngx-charts-circle-series
      ...
      [tooltipTemplate]="seriesTooltipTemplate"
      ...
        />
       <ng-template #seriesTooltipTemplate let-model="model">
          {{model.name | date: 'dd.MM.yyyy'}} &diams; {{model.value | durationHhmm}}
        </ng-template>
   
   <svgs.../>
</ngx-my-chart>
```
Second issue is still there, unchanged. The vertical line's tooltip would still not find a "defined" value, thus the NaN:NaN.
The durationHhmm pipe takes a number and returns its string equivalent in HH:mm format. 

![float_to_hh_mm_pipe_2](https://user-images.githubusercontent.com/23047676/43461565-1977107a-94d4-11e8-9113-853b7db37614.png)

```
<ngx-charts-line-chart id='chart' 
    ...
    ...
    ...
   >
    <ng-template #tooltipTemplate let-model="model">
          write your custom tooltip html here. the data is in model
    </ng-template>
 </ngx-charts-line-chart>
```

After having the code above implemented, the tooltip value was indeed overwritten ( only for the vertical line values, not when I hover over a data point ) and now shows nothing, i.e. the data binding fails.

![float_to_hh_mm_2](https://user-images.githubusercontent.com/23047676/43405075-4ee07b22-9419-11e8-8045-5bdd69e34171.png)

As pseudo code intended:

```
<ngx-my-chart 
   [options]="options_values"
   <svgs.../>
   
   <svg:g ngx-charts-tooltip-area
      ...
      [tooltipTemplate]="tooltipTemplate"
      ...
        />
       <ng-template #tooltipTemplate let-model="model">
          {{model.name}}Name: {{model.value}}Value
        </ng-template>
   
   <svgs.../>
</ngx-my-chart>
```
And in the component.ts:
```
 @ContentChild('tooltipTemplate') tooltipTemplate: TemplateRef<any>;
 @ContentChild('seriesTooltipTemplate') seriesTooltipTemplate: TemplateRef<any>;
```
</details>

Any help/suggestion is very welcome.

### **Sorry for posting in the wrong section.**

**Current behavior**
Works as intended.

**Expected behavior**
To display custom labels.

**Reproduction of the problem**
Link in the description above

**What is the motivation / use case for changing the behavior?**
Being able to customize tooltips' contents

**Please tell us about your environment:**
OS: Win 10 x64, IDE: Eclipse EE

* **ngx-charts version:** 3.0.2

* **Angular version:** 6.0.2

* **Browser:** [all]

* **Language:** [TypeScript 2.3.3]

**Comment by marjan-georgiev** [comment]
I replied to your stackoverflow comment. Here's a stackblitz that shows how to use tooltip templates:
https://stackblitz.com/edit/vertical-bar-chart-yvy5bx?file=app%2Fapp.component.ts

Also, any custom properties you assign are not passed down. There is an open issue around that.

**Comment by veselinodjov** [comment]
@marjan-georgiev Thank you for your answer.

> any custom properties you assign are not passed down

What about the returned string, text only? I managed to send desired data through double piping, format it as wished and send it back, but the returned text would be displayed on **one line**. 

```
<ng-template #tooltipTemplate let-model="model">
   {{model | json | durationHhmm}}
</ng-template> 
```

```
<ng-template #seriesTooltipTemplate let-model="model">
   {{model | json | durationSeriesHhmm}}   
</ng-template>
```

The **durationHhmm** and **durationSeriesHhmm** pipes return a string.

I tried:
- Adding `ngPreserveWhitespaces:true` to `<ng-template>` and/or `<svg:g ngx-charts-tooltip-area>` 
- CSS Rule: `ngx-tooltip-content` with  `white-space: pre-line !important;`
- `'\n'` (normal ticks), `\n` (backticks ES6), `<br/>`, `&#10;` (Line Feed), `&#13;` (Carriage return), also tried to render `<pre>` at some point but gave up.

For reference: https://github.com/angular/angular/issues/23970

![hhmm_pipe_1](https://user-images.githubusercontent.com/23047676/43525028-014c43da-95a1-11e8-8957-577bd7391187.png)

![hhmm_pipe_2](https://user-images.githubusercontent.com/23047676/43525230-84e76a58-95a1-11e8-826f-83c0cb55196b.png)

The **♦** and **|** symbols were manually added just for aesthetic look.

Wished output would be:

For a single point:

```
Parameter name
Date ♦ Value
```

For multiple params:
```
Param_1 name ♦ Date ♦ Value
Param_2 name ♦ Date ♦ Value
Param_n name ♦ Date ♦ Value
...
```
Also, any idea on how to pass multiple objects so they look the way they used to before I overwrote the tooltip's template?

![hhmm_pipe_3](https://user-images.githubusercontent.com/23047676/43526810-5ed22746-95a5-11e8-8839-d5b5c9226699.png)


**TL;DR** : Acquired custom label but with no style, a one-liner. When tried to manipulate it, nothing helped. I've tried several other rapidly googled techniques, but didn't add them to the list. They didn't help either.

**Comment by marjan-georgiev** [comment]
You can have HTML inside the template, so you can structure it however you want. Simply add a <br /> tag where you'd like a line break.

**Comment by veselinodjov** [comment]
Hmm, I get an error. A single hard coded `<br />` wouldn't solve the issue, but I even cannot use that little. Just for the sake of having an example:

```
<ng-template #tooltipTemplate let-model="model">
   {{model | json | durationHhmm}}<br />
</ng-template> 
```
The `<br />` produces:
```
Error: Uncaught (in promise): Error: Template parse errors:
':svg:br' is not a known element:
1. If ':svg:br' is an Angular component, then verify that it is part of this module.
2. To allow any element add 'NO_ERRORS_SCHEMA' to the '@NgModule.schemas' of this component. (">
        <ng-template #tooltipTemplate let-model="model">
          {{model | json | durationHhmm}}[ERROR ->]<br />
        </ng-template> 

"): ng:///AnalysisModule/CustomLineChartComponent.html@153:41
```

I'm new to the Angular scene. Still trying to get around simple things.

**Comment by marjan-georgiev** [comment]
I'm not sure why you're getting that. Are you defining that template inside of an svg tag? If you look at the example stackblitz I posted above, you can see it works with HTML tags.

Try reproducing it in a stackblitz and I can take a look.

**Comment by veselinodjov** [comment]
> Are you defining that template inside of an svg tag

Oh my... One of the templates, for the single point, was inside a svg, thus the error. 

### Thank you!

Actually, a single thank you won't do it. I see you are based in Berlin, as well. Just say what's your drink preference and when. 🥇

**Comment by marjan-georgiev** [comment]
A fellow Berliner! I occasionally attend the Angular and Data Visualization meetups here. If you're ever at those and see me, come say hi. 🍻 

Glad we resolved the problem.
